File: companies/Wise/253.meeting-rooms-ii.java
import java.util.*;

/**
 * Algorithm:
 * Sort meetings by start time. A min-heap stores current meeting end times.
 * Remove all rooms that have ended before the current meeting starts, then add
 * the current end time. The maximum heap size is the required rooms.
 *
 * Java data structures:
 * PriorityQueue<Integer> is the min-heap of end times.
 *
 * Complexity:
 * Time O(n log n), space O(n).
 */
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        if (intervals.length == 0) {
            return 0;
        }
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        int best = 0;
        for (int[] interval : intervals) {
            int start = interval[0];
            int end = interval[1];
            while (!heap.isEmpty() && heap.peek() <= start) {
                heap.poll();
            }
            heap.offer(end);
            best = Math.max(best, heap.size());
        }
        return best;
    }
}

File: companies/Hive/253.meeting-rooms-ii.py
"""
Approach: Sweep meetings by start time with a min-heap of end times.
Data structure: the heap stores the end time of every room currently in use; the smallest end time is the room that frees first.
Interview logic: before starting a meeting, pop all rooms that have ended. Then allocate the current meeting's room by pushing its end. The maximum heap size is the room count needed.
Complexity: O(n log n) time, O(n) space.
Tests and edge cases: empty intervals return 0; touching intervals like [1,2] and [2,3] reuse a room; fully overlapping intervals require n rooms.
"""
from __future__ import annotations
import heapq
from typing import List

# @lc code=start
import heapq
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0
        intervals.sort(key=lambda x: x[0])
        heap = []
        best = 0
        for start, end in intervals:
            while heap and heap[0] <= start:
                heapq.heappop(heap)
            heapq.heappush(heap, end)
            best = max(best, len(heap))
        return best
# @lc code=end
